functorch/functorch/csrc/BatchRulesLinearAlgebra.cpp

```diff
@@ -176,6 +182,181 @@ householder_product_batch_rule(const Tensor &input, c10::optional<int64_t> input
   return std::make_tuple(at::linalg_householder_product(input_, tau_), 0);
 }
 
+template <char const *op_name, typename A, A a, typename C>
+struct LinalgCheckMatrixUnaryRuleHelper;
+
+template <char const *op_name, typename F, F Func, typename A, typename... T>
+struct LinalgCheckMatrixUnaryRuleHelper<op_name, F, Func, typelist<A, T...>> {
+  static inline Tensor check_and_reshape_input(const Tensor& tensor, optional<int64_t> batch_dim) {
+    TORCH_CHECK(tensor.dim() > 2, op_name, ": The input tensor A must have at least 2 dimensions.");
+    return moveBatchDimToFront(tensor, batch_dim);
+  }
+
+  static oneOutput apply_one(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    return std::make_tuple(Func(tensor_, std::forward<T>(extra_args)...), 0);
+  }
+
+  static twoOutputs apply_two(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    const auto res = Func(tensor_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0);
+  }
+
+  static threeOutputs apply_three(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    const auto res = Func(tensor_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0, std::get<2>(res), 0);
+  }
+
+  static fourOutputs apply_four(
+      const Tensor& tensor,
+      optional<int64_t> batch_dim,
+      T... extra_args) {
+    const auto tensor_ = check_and_reshape_input(tensor, batch_dim);
+    const auto res = Func(tensor_, std::forward<T>(extra_args)...);
+    return std::make_tuple(std::get<0>(res), 0, std::get<1>(res), 0, std::get<2>(res), 0, std::get<3>(res), 0);
+  }
+};
+
+template <char const *op_name, typename A, A a, typename C>
+struct LinalgCheckMatrixBinaryRuleHelper;
+
+template <char const *op_name, typename F, F Func, typename A, typename B, typename... T>
+struct LinalgCheckMatrixBinaryRuleHelper<op_name, F, Func, typelist<A, B, T...>> {
+  static twoOutputs apply_two(
+      const Tensor& first, optional<int64_t> first_bdim,
+      const Tensor& second, optional<int64_t> second_bdim,
+      T... extra_args) {
+    TORCH_CHECK(!first_bdim.has_value() || first.dim() > 2,
+              op_name, ": The input tensor A must have at least 2 dimensions.");
+    TORCH_CHECK(!second_bdim.has_value() || second.dim() > 2,
+              op_name, ": The input tensor A must have at least 2 dimensions.");
```

Comment:
What exactly are these checks supposed to check for? that `first`, `second` are both logically matrices or batches of matrices?

If so we can just check `rankWithoutBatchDim(first, first_bdim) >= 2`.